ocaml/xapi/livepatch.ml

```diff
@@ -0,0 +1,279 @@
+
+let get_latest_livepatch lps =
+  List.map (fun (_, _, t_v, t_r) -> (t_v, t_r)) lps
+  |> Rpm.get_latest_version_release
+  |> function
+  | Some (v, r) ->
+      List.find_opt (fun (_, _, t_v, t_r) -> t_v = v && t_r = r) lps
+  | None ->
+      None
+
+module BuildId = struct
+  let one_byte =
+    let open Angstrom in
+    any_char >>= fun c ->
+    let code = Char.code c in
+    return (Printf.sprintf "%0x" code)
+
+  let all_bytes = Angstrom.many one_byte
+end
+
+module KernelLivePatch = struct
+  let get_running_livepatch () =
+    let r =
+      (* This expects the kernel livepatch module name should be built as:
+       * livepatch_<base_version>__<base_release>__<to_version>__<to_releaes>
+       * The "." in <base_version>, <base_release>, <to_version>, and <to_releaes>
+       * should be replaced with "_".
+       * For example:
+       * livepatch_4_19_19__8_0_20__4_19_19__8_0_22
+       *)
+      Re.Posix.compile_pat
+        {|^[ ]*livepatch_([^ \[]+)__([^ \[]+)__([^ \[]+)__([^ \[]+).*$|}
+    in
+    Helpers.call_script !Xapi_globs.kpatch_cmd ["list"]
+    |> Astring.String.cuts ~sep:"\n"
+    |> List.fold_left
+         (fun (acc, acc_flag) line ->
+           match
+             ( acc_flag
+             , Astring.String.is_prefix ~affix:"Installed patch modules:" line
+             )
+           with
+           | true, _ ->
+               (acc, true)
+           | false, true ->
+               (acc, true)
+           | false, false ->
+               (line :: acc, false)
+         )
+         ([], false)
+    |> (fun (x, _) -> x)
+    |> List.filter_map (fun line ->
+           let replace s =
+             Astring.String.cuts ~sep:"-" s |> Astring.String.concat ~sep:"."
+           in
+           match Re.exec_opt r line with
+           | Some groups ->
+               Some
+                 ( Re.Group.get groups 1 |> replace
+                 , Re.Group.get groups 2 |> replace
+                 , Re.Group.get groups 3 |> replace
+                 , Re.Group.get groups 4 |> replace
+                 )
+           | None ->
+               None
+       )
+    |> get_latest_livepatch
+
+  let align x =
+    (* The actual size is aligned with 4. *)
+    let open Int32 in
+    (if rem x 4l = 0l then x else add x (sub 4l (rem x 4l))) |> to_int
+
+  let rec section' () =
+    let open Angstrom in
+    at_end_of_input >>= function
+    | true ->
+        fail "Not found build id"
+    | false -> (
+        LE.any_int32 >>= fun name_size ->
+        LE.any_int32 >>= fun desc_size ->
+        LE.any_int32 >>= fun note_type ->
+        take (align name_size) >>= fun name ->
+        take (align desc_size) >>= fun desc ->
+        let name' = Astring.(String.trim ~drop:Char.Ascii.is_control name) in
+        match (note_type, desc_size, name') with
+        | 3l, 20l, "GNU" ->
+            return desc
+        | _ ->
+            section' ()
+      )
+
+  let section = section' ()
```

Comment:
Could we rename `section'` to `section` and get rid of this?

Reply:
I was not quite understand the error in that case:
`Error: This kind of expression is not allowed as right-hand side of 'let rec'`
So I added a `section`.